mono/metadata/monitor.c

```diff
@@ -1292,46 +1292,29 @@ ves_icall_System_Threading_Monitor_Monitor_pulse (MonoObject *obj)
 
 	mon = lock_word_get_inflated_lock (lw);
 
-	LOCK_DEBUG (g_message ("%s: (%d) %d threads waiting", __func__, mono_thread_info_get_small_id (), g_slist_length (mon->wait_list)));
+	LOCK_DEBUG (g_message ("%s: (%d) %d threads waiting", func, mono_thread_info_get_small_id (), g_slist_length (mon->wait_list)));
 
+again:
 	if (mon->wait_list != NULL) {
-		LOCK_DEBUG (g_message ("%s: (%d) signalling and dequeuing handle %p", __func__, mono_thread_info_get_small_id (), mon->wait_list->data));
-	
+		LOCK_DEBUG (g_message ("%s: (%d) signalling and dequeuing handle %p", func, mono_thread_info_get_small_id (), mon->wait_list->data));
+
 		mono_w32event_set (mon->wait_list->data);
 		mon->wait_list = g_slist_remove (mon->wait_list, mon->wait_list->data);
+		if (all)
+			goto again;
```

Comment:
can we use a loop instead of `goto`?

Reply:
Yes but it isn’t pretty. Something like do while all and not null.

Reply:
Or while true and break if null or not all.